File: engine/render/src/render/buffer.cpp
// Copyright 2020-2026 The Defold Foundation
// Copyright 2014-2020 King
// Copyright 2009-2014 Ragnar Svensson, Christian Murray
// Licensed under the Defold License version 1.0 (the "License"); you may not use
// this file except in compliance with the License.
//
// You may obtain a copy of the License, together with FAQs at
// https://www.defold.com/license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

#include "render_private.h"

namespace dmRender
{
    static HRenderBuffer NewRenderBuffer(dmGraphics::HContext graphics_context, RenderBufferType type)
    {
        switch(type)
        {
            case RENDER_BUFFER_TYPE_VERTEX_BUFFER: return (HRenderBuffer) dmGraphics::NewVertexBuffer(graphics_context, 0, 0x0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
            case RENDER_BUFFER_TYPE_INDEX_BUFFER:  return (HRenderBuffer) dmGraphics::NewIndexBuffer(graphics_context, 0, 0x0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
            default:break;
        }
        return (HRenderBuffer) 0;
    }

    static void DeleteRenderBuffer(HRenderBuffer buffer_to_remove, RenderBufferType type)
    {
        switch(type)
        {
            case RENDER_BUFFER_TYPE_VERTEX_BUFFER:
                dmGraphics::DeleteVertexBuffer((dmGraphics::HVertexBuffer) buffer_to_remove);
                break;
            case RENDER_BUFFER_TYPE_INDEX_BUFFER:
                dmGraphics::DeleteIndexBuffer((dmGraphics::HIndexBuffer) buffer_to_remove);
                break;
            default:break;
        }
    }

    static inline void CreateAndPush(HRenderContext render_context, HBufferedRenderBuffer buffer)
    {
        buffer->m_Buffers.Push(NewRenderBuffer(render_context->m_GraphicsContext, buffer->m_Type));
    }

    HBufferedRenderBuffer NewBufferedRenderBuffer(HRenderContext render_context, RenderBufferType type)
    {
        BufferedRenderBuffer* buffer = new BufferedRenderBuffer();
        memset(buffer, 0, sizeof(BufferedRenderBuffer));

        buffer->m_Type = type;
        buffer->m_Buffers.SetCapacity(1);

        RewindBuffer(render_context, buffer);
        CreateAndPush(render_context, buffer);

        return buffer;
    }

    void DeleteBufferedRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
    {
        if (!buffer)
            return;

        for (int i = 0; i < buffer->m_Buffers.Size(); ++i)
        {
            DeleteRenderBuffer(buffer->m_Buffers[i], buffer->m_Type);
        }
        delete buffer;
    }

    HRenderBuffer AddRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
    {
        if (!buffer)
            return 0;

        if (render_context->m_MultiBufferingRequired)
        {
            buffer->m_BufferIndex++;

            if (buffer->m_BufferIndex > 0)
            {
                if (buffer->m_Buffers.Full())
                {
                    buffer->m_Buffers.OffsetCapacity(4);
                }

                CreateAndPush(render_context, buffer);
            }
        }

        return buffer->m_Buffers[buffer->m_BufferIndex];
    }

    void SetBufferData(HRenderContext render_context, HBufferedRenderBuffer buffer, uint32_t size, void* data, dmGraphics::BufferUsage buffer_usage)
    {
        switch(buffer->m_Type)
        {
            case RENDER_BUFFER_TYPE_VERTEX_BUFFER:
            {
                dmGraphics::HVertexBuffer vbuf = (dmGraphics::HVertexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
                dmGraphics::SetVertexBufferData(vbuf, size, data, buffer_usage);
            } break;
            case RENDER_BUFFER_TYPE_INDEX_BUFFER:
            {
                dmGraphics::HIndexBuffer ibuf = (dmGraphics::HIndexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
                dmGraphics::SetIndexBufferData(ibuf, size, data, buffer_usage);
            } break;
            default:break;
        }
    }

    void SetBufferSubData(HRenderContext render_context, HBufferedRenderBuffer buffer, uint32_t offset, uint32_t size, void* data)
    {
        switch(buffer->m_Type)
        {
        case RENDER_BUFFER_TYPE_VERTEX_BUFFER:
        {
            dmGraphics::HVertexBuffer vbuf = (dmGraphics::HVertexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
            dmGraphics::SetVertexBufferSubData(vbuf, offset, size, data);
        } break;
        case RENDER_BUFFER_TYPE_INDEX_BUFFER:
        {
            dmGraphics::HIndexBuffer ibuf = (dmGraphics::HIndexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
            dmGraphics::SetIndexBufferSubData(ibuf, offset, size, data);
        } break;
        default:break;
        }
    }

    void TrimBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
    {
        if (!buffer)
            return;

        uint32_t n = buffer->m_Buffers.Size();
        uint32_t new_buffer_count = dmMath::Max(1, buffer->m_BufferIndex+1);

        for (int i = new_buffer_count; i < n; ++i)
        {
            DeleteRenderBuffer(buffer->m_Buffers[i], buffer->m_Type);
        }
        buffer->m_Buffers.SetSize(new_buffer_count);
    }

    HRenderBuffer GetBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
    {
        return buffer->m_Buffers[buffer->m_BufferIndex];
    }

    int32_t GetBufferIndex(HRenderContext render_context, HBufferedRenderBuffer buffer)
    {
        return buffer->m_BufferIndex;
    }

    void RewindBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
    {
        if (!buffer)
            return;
        buffer->m_BufferIndex = 0;
    }
}

File: editor/src/clj/editor/gl/buffer.clj
;; Copyright 2020-2025 The Defold Foundation
;; Copyright 2014-2020 King
;; Copyright 2009-2014 Ragnar Svensson, Christian Murray
;; Licensed under the Defold License version 1.0 (the "License"); you may not use
;; this file except in compliance with the License.
;;
;; You may obtain a copy of the License, together with FAQs at
;; https://www.defold.com/license
;;
;; Unless required by applicable law or agreed to in writing, software distributed
;; under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
;; specific language governing permissions and limitations under the License.

(ns editor.gl.buffer
  (:require [editor.buffers :as buffers]
            [editor.gl :as gl]
            [editor.gl.protocols :refer [GlBind]]
            [editor.graphics.types :as types]
            [editor.scene-cache :as scene-cache]
            [util.defonce :as defonce]
            [util.ensure :as ensure])
  (:import [com.jogamp.opengl GL2]
           [editor.buffers BufferData]
           [editor.graphics.types ElementType]
           [java.nio Buffer FloatBuffer IntBuffer ShortBuffer]))

(set! *warn-on-reflection* true)
(set! *unchecked-math* :warn-on-boxed)

(defonce/record ^:private GlBufferData
  [^BufferData buffer-data
   ^int gl-target
   ^int gl-usage])

(defn- data-type->gl-type
  ^long [data-type]
  (case data-type
    :type-byte GL2/GL_BYTE
    :type-unsigned-byte GL2/GL_UNSIGNED_BYTE
    :type-short GL2/GL_SHORT
    :type-unsigned-short GL2/GL_UNSIGNED_SHORT
    :type-int GL2/GL_INT
    :type-unsigned-int GL2/GL_UNSIGNED_INT
    :type-float GL2/GL_FLOAT))

(defn- target->gl-target
  ^long [target]
  (case target
    :array-buffer GL2/GL_ARRAY_BUFFER
    :element-array-buffer GL2/GL_ELEMENT_ARRAY_BUFFER))

(defn- usage->gl-usage
  ^long [usage]
  (case usage
    :static GL2/GL_STATIC_DRAW
    :dynamic GL2/GL_DYNAMIC_DRAW))

(defn- make-gl-buffer-data
  ^GlBufferData [^Buffer data target usage]
  (assert (or (.isDirect data) (.hasArray data)) "The Buffer must either be direct or backed by an array.")
  (let [buffer-data (buffers/make-buffer-data data)
        gl-target (target->gl-target target)
        gl-usage (usage->gl-usage usage)]
    (->GlBufferData buffer-data gl-target gl-usage)))

(defn- invalidate-gl-buffer-data
  ^GlBufferData [^GlBufferData gl-buffer-data]
  (->GlBufferData
    (buffers/invalidate-buffer-data (.-buffer-data gl-buffer-data))
    (.-gl-target gl-buffer-data)
    (.-gl-usage gl-buffer-data)))

(defn- request-gl-buffer!
  ^long [^GL2 gl request-id ^GlBufferData gl-buffer-data]
  (let [gl-buffer (scene-cache/request-object! ::gl-buffer request-id gl gl-buffer-data)]
    gl-buffer))

(defn- bind-gl-buffer!
  ^long [^GL2 gl request-id ^GlBufferData gl-buffer-data]
  (let [gl-target (.-gl-target gl-buffer-data)
        gl-buffer (request-gl-buffer! gl request-id gl-buffer-data)]
    (gl/gl-bind-buffer gl gl-target gl-buffer)
    gl-buffer))

(defn- unbind-gl-buffer!
  [^GL2 gl ^GlBufferData gl-buffer-data]
  (let [gl-target (.-gl-target gl-buffer-data)]
    (gl/gl-bind-buffer gl gl-target 0)))

(defonce/record BufferLifecycle
  [request-id
   ^GlBufferData gl-buffer-data
   ^ElementType buffer-element-type]

  GlBind
  (bind [_this gl _render-args]
    (bind-gl-buffer! gl request-id gl-buffer-data)
    nil)

  (unbind [_this gl _render-args]
    (unbind-gl-buffer! gl gl-buffer-data)))

(defn- make-buffer
  ^BufferLifecycle [request-id ^Buffer data ^ElementType buffer-element-type target usage]
  (ensure/argument request-id types/request-id?)
  (ensure/argument buffer-element-type types/element-type?)
  (let [gl-buffer-data (make-gl-buffer-data data target usage)]
    (->BufferLifecycle request-id gl-buffer-data buffer-element-type)))

(defn make-attribute-buffer
  ^BufferLifecycle [request-id ^FloatBuffer data vector-type usage]
  (if (instance? FloatBuffer data)
    (let [buffer-element-type (types/make-element-type vector-type :type-float false)]
      (make-buffer request-id data buffer-element-type :array-buffer usage))
    (throw (IllegalArgumentException. "data must be a FloatBuffer"))))

(defn attribute-buffer? [^BufferLifecycle value]
  (and (instance? BufferLifecycle value)
       (= GL2/GL_ARRAY_BUFFER (.-gl-target ^GlBufferData (.-gl-buffer-data value)))))

(defn make-index-buffer
  ^BufferLifecycle [request-id ^Buffer data usage]
  (let [data-type (condp instance? data
                    ShortBuffer :type-unsigned-short
                    IntBuffer :type-unsigned-int
                    (throw (IllegalArgumentException. "data must be either a ShortBuffer or an IntBuffer")))
        buffer-element-type (types/make-element-type :vector-type-scalar data-type false)]
    (make-buffer request-id data buffer-element-type :element-array-buffer usage)))

(defn index-buffer? [^BufferLifecycle value]
  (and (instance? BufferLifecycle value)
       (= GL2/GL_ELEMENT_ARRAY_BUFFER (.-gl-target ^GlBufferData (.-gl-buffer-data value)))))

(defn data
  ^Buffer [^BufferLifecycle buffer-lifecycle]
  (.-data ^BufferData (.-buffer-data ^GlBufferData (.-gl-buffer-data buffer-lifecycle))))

(defn invalidate
  ^BufferLifecycle [^BufferLifecycle buffer-lifecycle]
  (->BufferLifecycle (.-request-id buffer-lifecycle)
                     (invalidate-gl-buffer-data (.-gl-buffer-data buffer-lifecycle))
                     (.-buffer-element-type buffer-lifecycle)))

(defn vertex-count
  ^long [^BufferLifecycle buffer-lifecycle]
  (if (nil? buffer-lifecycle)
    0
    (let [buffer (data buffer-lifecycle)
          item-count (buffers/item-count buffer)]
      (if (zero? item-count)
        0
        (let [buffer-element-type ^ElementType (.-buffer-element-type buffer-lifecycle)
              vector-type (.-vector-type buffer-element-type)
              component-count (types/vector-type-component-count vector-type)]
          (quot item-count component-count))))))

(defn assign-attribute!
  [^BufferLifecycle buffer-lifecycle ^GL2 gl ^long base-location]
  ;; Note: Assumes both the BufferLifecycle and the ShaderLifecycle are bound.
  (let [buffer-element-type ^ElementType (.-buffer-element-type buffer-lifecycle)
        vector-type (.-vector-type buffer-element-type)
        data-type (.-data-type buffer-element-type)
        normalize (.-normalize buffer-element-type)
        items-per-vertex (types/vector-type-component-count vector-type)
        item-gl-type (data-type->gl-type data-type)
        item-byte-size (types/data-type-byte-size data-type)
        vertex-byte-size (* items-per-vertex item-byte-size)
        row-column-count (types/vector-type-row-column-count vector-type)]
    (if (neg? row-column-count)
      (do
        (gl/gl-enable-vertex-attrib-array gl base-location)
        (gl/gl-vertex-attrib-pointer gl base-location items-per-vertex item-gl-type normalize vertex-byte-size 0))
      (loop [column-index 0]
        (when (< column-index row-column-count)
          (let [location (+ base-location column-index)
                byte-offset (* column-index row-column-count item-byte-size)]
            (gl/gl-enable-vertex-attrib-array gl location)
            (gl/gl-vertex-attrib-pointer gl location row-column-count item-gl-type normalize vertex-byte-size byte-offset)
            (recur (inc column-index))))))))

(defn clear-attribute!
  [^BufferLifecycle buffer-lifecycle ^GL2 gl ^long base-location]
  (let [buffer-element-type ^ElementType (.-buffer-element-type buffer-lifecycle)
        vector-type (.-vector-type buffer-element-type)
        attribute-count (types/vector-type-attribute-count vector-type)]
    (gl/clear-attributes! gl base-location attribute-count)))

(defn- update-gl-buffer!
  [^GL2 gl ^long gl-buffer ^GlBufferData gl-buffer-data]
  (let [gl-target (.-gl-target gl-buffer-data)
        gl-usage (.-gl-usage gl-buffer-data)
        buffer-data ^BufferData (.-buffer-data gl-buffer-data)
        data (.-data buffer-data)
        data-byte-size (buffers/total-byte-size data)]
    (assert (buffers/flipped? data) "data buffer must be flipped before use")
    (gl/gl-bind-buffer gl gl-target gl-buffer)
    (gl/gl-buffer-data gl gl-target data-byte-size data gl-usage)
    gl-buffer))

(defn- create-gl-buffer!
  [^GL2 gl ^GlBufferData gl-buffer-data]
  (let [gl-buffer (gl/gl-gen-buffer gl)
        gl-buffer (update-gl-buffer! gl gl-buffer gl-buffer-data)]
    gl-buffer))

(defn- destroy-gl-buffers!
  [^GL2 gl gl-buffers _gl-buffer-datas]
  (gl/gl-delete-buffers gl gl-buffers))

(scene-cache/register-object-cache! ::gl-buffer create-gl-buffer! update-gl-buffer! destroy-gl-buffers!)
